feat: add validator SkelBindingAPIAppliedChecker

beersandrew opened the pull request:
### Description of Change(s)
add validator SkelBindingAPIAppliedChecker
- added a validation check for when a skel binding property is present but the skel binding api is not applied
- added a validation check for when a skel binding api is applied to a prim that is not parented by a skel root
- added a unit test

### Fixes Issue(s)
- https://github.com/orgs/PixarAnimationStudios/projects/3/views/1?pane=issue&itemId=65425045
- 
<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: add validator SkelBindingAPIAppliedChecker

- added a validation check for when a skel binding property is present but the skel binding api is not applied
- added a validation check for when a skel binding api is applied to a prim that is not parented by a skel root
- added a unit test

jesschimein commented:
Filed as internal issue #USD-9849

jesschimein commented:
/AzurePipelines run

dgovil commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+    TF_AXIOM(validator);
+
+    // Create Stage and mesh with a skel binding property
+    UsdStageRefPtr usdStage = UsdStage::CreateNew("/Users/andrewbeers/test.usda");
```
Nit: I'm sure you'll remove it but good to just write it to a temp test location instead before it's merged in, and/or just keep it in memory if you don't need to write it out.

dgovil commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+
+    UsdValidationErrorVector errors = validator->Validate(mesh.GetPrim());
+
+    // test and verify error for not having a skel binding api applied
```
Smallest of nitpicks: Comment style should be sentence case just to match the rest of the comments.

dgovil commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+    TF_AXIOM(errors[0].GetSites()[0].IsPrim());
+    TF_AXIOM(errors[0].GetSites()[0].GetPrim().GetPath() ==
+             SdfPath("/SkelRoot/Mesh"));
+    expectedErrorMsg = "UsdSkelBindingAPI applied on prim: </SkelRoot/Mesh>, " \
```
Style nit: You can put the string inside parenthesis and avoid escaping the line break. It's a little cleaner and less fragile if someone forgets in the future.

beersandrew replied:
didn't know about this! thanks!

dgovil commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,90 @@
+                                    UsdValidationErrorSite(usdPrim.GetStage(),
+                                                           usdPrim.GetPath())
+                            },
+                            TfStringPrintf("Found a UsdSkelBinding property (%s), " \
```
Perhaps a question for @tallytalwar , but are we always printing out the warning messages during the validator, or do we collect it and then print it at the end?

dgovil commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,90 @@
+
+        for (const TfToken &primToken : primPropertyNames){
+            for (const TfToken &skelToken : skelPropertyNames){
+                if (skelToken == primToken){
```
Nit: If you invert this comparison + use a continue,  you can unnest the contents of the check. Ever so slightly improves readability.

dgovil commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,90 @@
+    else {
+        if (usdPrim.GetTypeName() != UsdSkelTokens->SkelRoot) {
+            UsdPrim parentPrim = usdPrim.GetParent();
+            while (!parentPrim.IsPseudoRoot()) {
```
I think you should also check if the parentPrim isn't null, just in case you get a bad prim here.
`while (parentPrim && !parentPrim.IsPseudoRoot())`

dgovil commented:
LGTM! other than a couple small notes, most of which are just minor style ones.

beersandrew committed:
fix: fix issues based on PR comments

- update comments to be in sentence form
- remove local path, update to an in memory stage
- remove \ from multi-line strings and wrap them in parenthesis instead
- reduce nesting in validation code
- check parentPrim for null before checking it's type

beersandrew commented:
> LGTM! other than a couple small notes, most of which are just minor style ones.

Made the suggested changes. Thanks so much for the review & ideas.

tallytalwar commented:
Just curious why is this a draft PR?

tallytalwar commented on pxr/usd/usdSkel/CMakeLists.txt:
```diff
@@ -118,11 +132,6 @@ pxr_install_test_dir(
     DEST testUsdSkelCache
 )
 
-pxr_install_test_dir(
```
Did you intend to remove this test?

beersandrew replied:
this was not intentional, re-added.

tallytalwar commented on pxr/usd/usdSkel/CMakeLists.txt:
```diff
@@ -93,6 +97,16 @@ pxr_library(usdSkel
+
+pxr_build_test(testUsdSkelValidators
+        LIBRARIES
+        sdr
```
I am not sure if sdr is needed as a dependency here. Can you remove this please? Thanks

beersandrew replied:
removed, thanks

tallytalwar commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+
+#include "pxr/pxr.h"
+#include "pxr/base/tf/token.h"
+#include "pxr/usd/sdr/registry.h"
```
Can you remove these extra headers? Thanks

beersandrew replied:
updated

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,91 @@
+// https://openusd.org/license.
+//
+
+#include "pxr/usd/sdr/shaderProperty.h"
```
Similarly sdr related headers shouldn't be required in usdSkel.

So this one and the sdr/registry.h below.

beersandrew replied:
removed

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,91 @@
+
+    if (!usdPrim.HasAPI<UsdSkelBindingAPI>()){
+        UsdSchemaRegistry& usdSchemaRegistry = UsdSchemaRegistry::GetInstance();
+        std::unique_ptr<UsdPrimDefinition> primDef = usdSchemaRegistry.BuildComposedPrimDefinition(TfToken(), {TfToken("SkelBindingAPI")});
+
+        std::vector<TfToken> skelPropertyNames = primDef->GetPropertyNames();
```
You can clean this further.

I would recommend something along these lines:

```cpp
static const std::vector<TfToken> skelPropertyNames = []() {
    UsdSchemaRegistry& usdSchemaRegistry = UsdSchemaRegistry::GetInstance();
    std::unique_ptr<UsdPrimDefinition> primDef = usdSchemaRegistry.BuildComposedPrimDefinition(TfToken(), {TfToken("SkelBindingAPI")});
        return primDef->GetPropertyNames();
    }();
```

This skelProperties are not repeatedly computed and can be saved in the static variable here.

beersandrew replied:
great suggestion, updated

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,91 @@
+        for (const TfToken &primToken : primPropertyNames){
+            for (const TfToken &skelToken : skelPropertyNames){
+                if (skelToken != primToken)
+                    continue;
```
Most of openUSD code follows a convention of surrounding if blocks with {}, even for single line blocks.

beersandrew replied:
fixed

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,91 @@
+        std::vector<TfToken> skelPropertyNames = primDef->GetPropertyNames();
+        std::vector<TfToken> primPropertyNames = usdPrim.GetPropertyNames();
+
+        for (const TfToken &primToken : primPropertyNames){
+            for (const TfToken &skelToken : skelPropertyNames){
+                if (skelToken != primToken)
+                    continue;
```
Also maybe make skelPropertyNames in unordered_set of Tokens, then you can only iterate on primProperties and do a find of the primProperty in the set, if any is found in the set, then you can report the validation error below and break the for loop! No need to report multiple errors in this case I think.

I think this way it will be more performant. Let me know what you think?

beersandrew replied:
i think this is a great idea, implemented.

tallytalwar commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+int
+main()
+{
+    TestUsdSkelBindingAPIChecker();
```
Can you also add that subset UsdSkelValidator keywords test here (similar to what you did for UsdGeom / UsdShade)?

beersandrew replied:
added

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,91 @@
+                                                       usdPrim.GetPath())
+                        },
+                        TfStringPrintf(("Found a UsdSkelBinding property (%s), "
+                            "but no SkelBindingAPI applied on the prim <%s>.(fails 'SkelBindingAPIAppliedChecker')"),
```
I will also recommend to remove the test name from the error string. Validator which resulted in this error will automatically be embedded in the Error (https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usd/validator.cpp#L61).

beersandrew replied:
removed

tallytalwar commented on pxr/usd/usdSkel/plugInfo.json:
```diff
@@ -69,6 +69,14 @@
                     }
+                },
+                "Validators": {
+                    "SkelBindingApiAppliedChecker": {
```
I think in the 2 validators I introduced in the dev branch, I was trying not to use "Checker" in the name. So

`SkelBindingAPIApplicationValidator` or `SkelBindingAPIApplicationCompliance` ?

cc. @dgovil if you have opinions on this?

beersandrew replied:
I did not change this yet, as I'll wait for some agreement (same issue in the MaterialBindingAPI PR). 

My suggestions: (or any combinations of:
 `<APISchema>` + `<Existence | Presence>` + `<Validator | Verifier | Checker>`
 )
- `SkelBindingAPIExistenceVerifier`
- `SkelBindingAPIPresenceValidator`
- ...

`Application` here makes me think of a program / app, rather than being applied, but I'm happy to move forward with any of the options.

dgovil replied:
Since the schemas are an appliedAPI schema, I kind of like  `SkelBindingAPIAppliedValidator`

tallytalwar replied:
I am fine with that. Will also be renaming the StageMetadataChecker ones I have already synced in to be called StageMetadataValidator. Thanks both.

tallytalwar commented on pxr/usd/usdSkel/testenv/testUsdSkelValidators.cpp:
```diff
@@ -0,0 +1,88 @@
+    errors = validator->Validate(mesh.GetPrim());
+
+    // Verify all errors are gone
+    TF_AXIOM(errors.size() == 0);
```
Very nice test!!!

tallytalwar reviewed (commented):
Hey @beersandrew 

Thanks for doing this. Left a few suggestions for you to consider. Let me know if you want to talk through any of these or more here or on slack.

beersandrew committed:
fix: address PR comments

- reenabled testUsdSkelRoot, accidentally removed in a previous commit
- removed sdr dependency from skel test
- removed unused headers
- added TestUsdSkelValidators test to verify validator existence
- removed SkelBindingApiChecker name from error messages
- refactored how skelPropertyNames are collected, stored, and checked against

beersandrew committed:
refactor: break out once an error is found

- no need to catch multiple errors here

beersandrew committed:
fix: rename checker to validator

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,89 @@
+    UsdValidationErrorVector errors;
+
+    if (!usdPrim.HasAPI<UsdSkelBindingAPI>()){
+        std::vector<TfToken> primPropertyNames = usdPrim.GetPropertyNames();
```
nitpick: Its a good practice to declare variables just before use. So I would move  `primPropertyNames` below `skelPropertyNames`.

Also you can make this a `const vector`.

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,89 @@
+        std::vector<TfToken> primPropertyNames = usdPrim.GetPropertyNames();
+        static const std::unordered_set<TfToken, TfHash> skelPropertyNames = []() {
+            UsdSchemaRegistry& usdSchemaRegistry = UsdSchemaRegistry::GetInstance();
+            std::unique_ptr<UsdPrimDefinition> primDef = usdSchemaRegistry.BuildComposedPrimDefinition(TfToken(), {TfToken("SkelBindingAPI")});
```
Ahh apologies, I should have noticed this before, you can use `UsdSkelTokens->SkelBindingAPI`, instead of `TfToken("SkelBindingAPI")`. Such tokens are often generated along with the schema.

https://github.com/PixarAnimationStudios/OpenUSD/blob/release/pxr/usd/usdSkel/tokens.cpp#L55

tallytalwar commented on pxr/usd/usdSkel/validators.cpp:
```diff
@@ -0,0 +1,89 @@
+        static const std::unordered_set<TfToken, TfHash> skelPropertyNames = []() {
+            UsdSchemaRegistry& usdSchemaRegistry = UsdSchemaRegistry::GetInstance();
+            std::unique_ptr<UsdPrimDefinition> primDef = usdSchemaRegistry.BuildComposedPrimDefinition(TfToken(), {TfToken("SkelBindingAPI")});
+            std::vector<TfToken> vectorOfNames = primDef->GetPropertyNames();
```
`const vector`

beersandrew committed:
fix: addressed PR comments

- switched from using TfToken("SkelBindingAPI") to UsdSkelTokens->SkelBindingAPI
- use const vectors where possible
- refactoring / reordering of code for clarity

tallytalwar reviewed (approved):
Thanks a bunch @beersandrew. Will sync this in.

jesschimein commented:
/AzurePipelines run
